src/lamp_py/ingestion/light_rail_gps.py

```diff
@@ -0,0 +1,225 @@
+import os
+from typing import (
+    List,
+    Optional,
+    Tuple,
+)
+from threading import current_thread
+from concurrent.futures import ThreadPoolExecutor
+
+from pyarrow import fs
+import polars as pl
+
+from lamp_py.aws.s3 import (
+    move_s3_objects,
+    file_list_from_s3,
+    object_exists,
+)
+from lamp_py.runtime_utils.process_logger import ProcessLogger
+
+from lamp_py.ingestion.utils import DEFAULT_S3_PREFIX
+
+raw_gps_schema = pl.Schema(
+    {
+        "serial_number": pl.String(),
+        "speed": pl.Float64(),
+        "date": pl.Date(),
+        "updated_at": pl.String(),
+        "bearing": pl.Float64(),
+        "latitude": pl.String(),
+        "longitude": pl.String(),
+        "car": pl.String(),
+    }
+)
+
+
+def thread_init(file: str) -> None:
+    """
+    initialize the filesystem in each read thread
+
+    :param file: example file to determine file system type
+    """
+    thread_data = current_thread()
+    if file.startswith("s3://"):
+        thread_data.__dict__["file_system"] = fs.S3FileSystem()
+    else:
+        thread_data.__dict__["file_system"] = fs.LocalFileSystem()
+
+
+def thread_gps_to_frame(path: str) -> Tuple[Optional[pl.DataFrame], str]:
+    """
+    gzip to dataframe converter function meant to be run in ThreadPool
+
+    :param path: path to gzip that will be converted to polars dataframe
+    """
+    file_system = current_thread().__dict__["file_system"]
+    path = path.replace("s3://", "")
+
+    logger = ProcessLogger(process_name="light_rail_gps_to_frame", path=path)
+    logger.log_start()
+
+    try:
+        with file_system.open_input_stream(path, compression="gzip") as f:
+            df = (
+                pl.read_json(f.read())
+                .transpose(
+                    include_header=True,
+                    header_name="serial_number",
+                    column_names=("data",),
+                )
+                .select(
+                    pl.col("serial_number").cast(pl.String),
+                    pl.col("data").struct.field("speed").cast(pl.Float64),
+                    (
+                        pl.col("data")
+                        .struct.field("updated_at")
+                        .str.slice(0, length=10)
+                        .str.to_date()
+                        .alias("date")
+                    ),
+                    pl.col("data").struct.field("updated_at").cast(pl.String),
+                    pl.col("data").struct.field("bearing").cast(pl.Float64),
+                    pl.col("data").struct.field("latitude").cast(pl.String),
+                    pl.col("data").struct.field("longitude").cast(pl.String),
+                    pl.col("data").struct.field("car").cast(pl.String),
+                )
+            )
```

Comment:
this seems to be a nicer way to do this than using pyarrow.

Reply:
Yeah, I also think most conversions between pyarrow and polars are zero-copy operations. The only tricky thing is making sure we're handling any type conversions correctly.